inverse/beamformer_lcmv.m

```diff
@@ -191,143 +207,189 @@
+      dat   = dip.subspace{i} * dat_pre_subspace;
+      C     = dip.subspace{i} * C_pre_subspace * dip.subspace{i}';
+      invC  = ft_inv(dip.subspace{i} * C_pre_subspace * dip.subspace{i}', invopt{:});
+    elseif ~isempty(subspace)
+      % do subspace projection of the forward model only
+      lforig = lf;
+      lf     = subspace * lf;
     
-    % according to Kensuke's paper, the eigenspace bf boils down to projecting
-    % the 'traditional' filter onto the subspace
-    % spanned by the first k eigenvectors [u,s,v] = svd(Cy); filt = ESES*filt; 
-    % ESES = u(:,1:k)*u(:,1:k)';
-    % however, even though it seems that the shape of the filter is identical to
-    % the shape it is obtained with the following code, the w*lf=I does not hold.
-  end
+      % according to Kensuke's paper, the eigenspace bf boils down to projecting
+      % the 'traditional' filter onto the subspace spanned by the first k eigenvectors 
+      % [u,s,v] = svd(C); filt = ESES*filt; ESES = u(:,1:k)*u(:,1:k)';
+      % however, even though it seems that the shape of the filter is identical to
+      % the shape it is obtained with the following code, the w*lf=I does not hold.
+    end
+
+    if fixedori
+      switch(weightnorm)
+        case {'unitnoisegain','nai'}
+          % optimal orientation calculation for unit-noise gain beamformer,
+          % (also applies nai weightnorm constraint), based on equation 4.47 from Sekihara & Nagarajan (2008)
+          % the following is a reformulation of the generalized eigenproblem 
+          [vv, dd]      = eig(pinv(lf' * invC_squared *lf)*(lf' * invC *lf));
+          [dum, maxeig] = max(diag(dd));
+          unitnoiseori  = vv(:,maxeig);
+          lf            = lf * unitnoiseori;
+          if hassubspace, lforig = lforig * unitnoiseori; end
+          dipout.ori{i} = unitnoiseori;
+
+        case 'arraygain'
+          % optimal orientation calculation for array-gain beamformer, Sekihara & Nagarajan eqn. 4.44
+          [vv, dd]      = eig(pinv(lf' * invC *lf)*(lf' * lf));
+          [dum, maxeig] = max(diag(dd));
+          arraygainori  = vv(:,maxeig);
+          lf            = lf * arraygainori;
+          if hassubspace, lforig = lforig * arraygainori; end
+          dipout.ori{i} = arraygainori;
+ 
+        otherwise
+          % compute the leadfield for the optimal dipole orientation that maximizes spatial filter output
+          % subsequently the leadfield for only that dipole orientation will be used for the final filter computation
+          % filt = pinv(lf' * invC * lf) * lf' * invC;
+          % [u, s, v] = svd(real(filt * C * ctranspose(filt)));
+          % in this step the filter computation is not necessary, use the quick way to compute the voxel level covariance (cf. van Veen 1997)
+          [u, s, v] = svd(real(pinv(lf' * invC *lf)));
+          maxpowori = u(:,1);
+          eta = s(1,1)./s(2,2); % ratio between the first and second singular valu
```

Comment:
This is new, no? Just curious whether this will perform more stable? (Taking the ratio).
Also, you are missing an "e" in the comment (more nitpicks!).

Reply:
You mean the 'eta' variable? This is just a ratio that is not used as an activity estimate, but can be inspected to get a feel for how justified it is to assume a fixed orientation at the location of interest. The closer to 1, the more arbitrary this assumption is.

Reply:
Ah, you renamed the actual orientation to `maxpowori`, I see. But interesting, you can basically plot that throughout the brain and see how good/bad it is to force optimal orientation?

Reply:
yes indeed something like that. I have fixed the typo and also allow for an 'eta' computation for the other fixed orientation computation methods.